### PR TITLE
Allow to subscribe to events on a specfic netns

### LINKED ISSUE #141: Subscribe functions should be exposed as methods of handles
I have a need to subscribe to routes and links in a different namespace, exposing the subscribe functions as methods of a netlink handle would be very helpful.

### PR DESCRIPTION
**Pull request by aboch**
Fixes #141 

Tests would be added, but want to get feedback first

Signed-off-by: Alessandro Boch <aboch@docker.com>

### CODE REVIEW

File: nl/nl_linux.go [hunk 1]
```diff
@@ -332,24 +332,47 @@ func getNetlinkSocket(protocol int) (*NetlinkSocket, error) {
...
+	}
+	defer c()
+	return getNetlinkSocket(protocol)
+}
 
+// ExecteInNetns sets execution of the code following this call to the
+// network namespace newNs, then moves the thread back to curNs if open,
+// otherwise to the current netns at the time the function was invoked
+// TODO: his function probably belongs to netns pkg. it could even be a method of NsHandle
+func ExecteInNetns(newNs, curNs netns.NsHandle) (c func(), err error) {
```
Comment (vishvananda) [inline nl/nl_linux.go]: s/Execte/Execute s/his/this/

File: nl/nl_linux.go [hunk 2]
```diff
@@ -332,24 +332,47 @@ func getNetlinkSocket(protocol int) (*NetlinkSocket, error) {
...
 		if !curNs.IsOpen() {
 			if curNs, err = netns.Get(); err != nil {
-				return nil, fmt.Errorf("could not get current namespace while creating netlink socket: %v", err)
+				curNs.Close()
+				return c, fmt.Errorf("could not get current namespace while creating netlink socket: %v", err)
 			}
-			defer curNs.Close()
+			a = curNs.Close
 		}
 		if err := netns.Set(newNs); err != nil {
```
Comment (vishvananda) [inline nl/nl_linux.go]: i think you need a curNs.Close() here since callers don't call c on error

File: route_linux.go [hunk 3]
```diff
@@ -425,6 +426,20 @@ func RouteSubscribe(ch chan<- RouteUpdate, done <-chan struct{}) error {
 	if err != nil {
 		return err
 	}
+	return routeSubscribe(s, ch, done)
+}
+
+// AddrSubscribeAt works like RouteSubscribe plus it allows the caller
```
Comment (vishvananda) [inline route_linux.go]: s/Addr/Route

**Comment by vishvananda** [comment]
looks reasonable. A few typos and one bug inline

File: nl/nl_linux.go (hunk 1 above)
Comment (aboch) [inline nl/nl_linux.go]: Will change

File: nl/nl_linux.go (hunk 2 above)
Comment (aboch) [inline nl/nl_linux.go]: True. Thanks for catching this.

File: route_linux.go (hunk 3 above)
Comment (aboch) [inline route_linux.go]: Thanks, will fix

**Comment by aboch** [comment]
Thanks @vishvananda , in the meanwhile I will do more tests offline.

File: addr_linux.go [hunk 4]
```diff
@@ -196,6 +197,20 @@ func AddrSubscribe(ch chan<- AddrUpdate, done <-chan struct{}) error {
 	if err != nil {
 		return err
 	}
+	return addrSubscribe(s, ch, done)
+}
+
+// AddrSubscribeAt works like AddrSubscribe plus it allows the caller
+// to choose the network namespace in which to subscribe (ns).
+func AddrSubscribeAt(ns netns.NsHandle, ch chan<- AddrUpdate, done <-chan struct{}) error {
+	s, err := nl.SubscribeAt(ns, netns.None(), syscall.NETLINK_ROUTE, syscall.RTNLGRP_LINK)
```
Comment (aboch) [inline addr_linux.go]: Wrong params

File: route_linux.go [hunk 5]
```diff
@@ -425,6 +426,20 @@ func RouteSubscribe(ch chan<- RouteUpdate, done <-chan struct{}) error {
 	if err != nil {
 		return err
 	}
+	return routeSubscribe(s, ch, done)
+}
+
+// RouteSubscribeAt works like RouteSubscribe plus it allows the caller
+// to choose the network namespace in which to subscribe (ns).
+func RouteSubscribeAt(ns netns.NsHandle, ch chan<- RouteUpdate, done <-chan struct{}) error {
+	s, err := nl.SubscribeAt(ns, netns.None(), syscall.NETLINK_ROUTE, syscall.RTNLGRP_LINK)
```
Comment (aboch) [inline route_linux.go]: Wrong params

**Comment by aboch** [comment]
@vishvananda Added few tests. PTAL when you get a chance. Thanks.

**Pull request merged**